### ISSUE TITLE
Back slashes added?

### ISSUE THREAD

**Issue by kevinSuttle** [pascalduez/postcss-apply#11]
Any ideas on what could be causing this? I'm not sure it's this plugin, to be honest. 

<img width="1151" alt="screen shot 2016-08-23 at 12 03 22 pm" src="https://cloud.githubusercontent.com/assets/95672/17899636/ea84c456-6929-11e6-9411-2fad07be3c31.png">

``` js
 postcss: function () {
      return [
        require('postcss-import')({
          addDependencyTo: webpack,
        }),
        customProperties({
          preserve: true,
        }),
        cssApply,
        autoprefixer,
      ]
    },
```

**Comment by kevinSuttle** [comment]
I even removed `:root` and still the back slashes were added. 

<img width="667" alt="screen shot 2016-08-23 at 12 30 51 pm" src="https://cloud.githubusercontent.com/assets/95672/17900488/8be817c8-692d-11e6-8045-c10e5ac4ba24.png">

**Comment by kevinSuttle** [comment]
There seems to be some issue with colons as well. When I don't add them to the "mixins" declared under `:root`, the rules below `:root` are not eval'd.

**Comment by pascalduez** [comment]
Hi,

are you sure the `@apply` plugin is executed actually? 
Because property sets declarations are supposed to be removed from final output.  

Or most probably it's because the slash is added to the declaration before the plugin runs, which breaks its "detection", so nothing happens.

I would advise to always add colons at the end of the declaration.
I will add more tests for this.  


Can you set-up a small repo to be able to reproduce/debug?

**Comment by kevinSuttle** [comment]
The plugin is running, because the styles are being applied. Let me see if I can set up a small test repo.

**Comment by kevinSuttle** [comment]
@pascalduez Here you go: https://cloudup.com/cLsnx9RtW82 See README for repro steps.

**Comment by kevinSuttle** [comment]
> The plugin is running, because the styles are being applied. 

Odd. I noticed that they're being applied in Chrome, but not Safari. 
<img width="633" alt="screen shot 2016-08-24 at 2 24 54 pm" src="https://cloud.githubusercontent.com/assets/95672/17942723/93f455ce-6a06-11e6-8a52-df2830ee7479.png">

**Comment by kevinSuttle** [comment]
Yeah, something's up for sure, because changing the buttons.css file doesn't change anything in Chrome, even though Webpack is "rebuilding".

**Comment by kevinSuttle** [comment]
Maybe it's this? 

https://github.com/postcss/postcss-custom-properties/issues/48

**Comment by pascalduez** [comment]
@kevinSuttle Is this issue still relevant?

**Comment by pascalduez** [comment]
I tried fiddling with the provided repo for an hour or so, I could not get any progress. 
Looks like the postcss plugins are not even applied/called.
By default storybook will use its own webpack config, so no surprise the plugins aren't used. 
I added a [custom webpack config for storybook](https://getstorybook.io/docs/configurations/custom-webpack-config) but same result.

Next step would be to try without storybook, just "plain" webpack and an index.html.

I'm using cssnext on a projet with webpack  n' stuff and have no issue BTW.

**Comment by kevinSuttle** [comment]
@pascalduez I haven't seen it reproduced. Thanks for checking into it. I appreciate you taking the time.

**Comment by kevinSuttle** [comment]
Seeing this again now.
I'm now using:

https://github.com/sebastian-software/postcss-smart-import
https://github.com/postcss/postcss-custom-properties/

```css
:root {
  \--ThemeButtonStyles: {
    @apply --ThemeAnimationStyles;
...
}
```

```js
postcss: function() {
      return [
        atImport({
           addDependencyTo: webpack,
           path: './src/Index.css',
           plugins: [
              customProperties({preserve: true}),
              cssApply,
              autoprefixer
          ]
        }),
      ]
    },
```

**Comment by pascalduez** [comment]
This is concerning enough, even if Webpack 2 is still in beta. Re-opening.

**Comment by kevinSuttle** [comment]
Let me elaborate on the structure:

**./Index.css**
```css
@import "./Themes/Vars.css";
@import "./Themes/Colors/Kratos.css";
@import "./Themes/Colors/Prometheus.css";
@import "./Themes/Mixins.css";
```

```shell
WARNING in ./~/css-loader!./~/postcss-loader!./src/Index.css
postcss-custom-properties: /Users/kevinSuttle/Code/platform-ui-primitives/src/Themes/Mixins.css:15:5: variable '--PrimaryThemeColor' is undefined and used without a fallback
```
_[repeats]_

This is inaccurate. --PrimaryThemeColor is defined in the 2 color files listed _before_ Mixins.css.
BUT, something is adding these slashes before they can get there.

**Comment by kevinSuttle** [comment]
Confirmed. Removing that `:root` declaration in Button/Buttons.css removed that `\`, but the warnings still appear. Separate issue it seems.

**Comment by kevinSuttle** [comment]
Hm. The top-level styles are getting added last, which some of it, but not all. 

<img width="831" alt="screen shot 2016-09-20 at 3 50 52 pm" src="https://cloud.githubusercontent.com/assets/95672/18686501/2047fea6-7f4a-11e6-9583-a12a60d43ede.png">

**Comment by kevinSuttle** [comment]
I did this in my top-level npm module export:

```
require("./Themes/Vars.css");
require("./Themes/Colors/Kratos.css");
require("./Themes/Colors/Prometheus.css");
require("./Themes/Mixins.css");
```

And got the slashes back on top-level `:root` declarations. 
I removed the smart-import plugin and the slashes went away. 

No warnings, no errors, no slashes.

**Comment by Memoyr** [comment]
Hi , I got the same issue with the plugin, it was adding backslashes when I try to import  custom property sets from an external file. 
I aim to keep all of my sets in separate file as my project is  build with multiple CSS files (files per component).
I succeed to import my file and use custom property sets by using postcss-apply in conjonction of postcss-import.
There's 2 downside. 
- I have to import the file into my current file anytime I want to use a custom property set.
- A waring is triggered saying "No custom properties set declared " , even though  the styles are actually being applied. 

I there a better way to achieve this, without those two downside ?
Thx

part of my webpack config : 
`
var pcssNesting       = require('postcss-nesting')
var pcssNext          = require('postcss-cssnext')
var cssApply          = require('postcss-apply')
var pcssCustomProps   = require('postcss-custom-properties')
var pcssReporter      = require('postcss-reporter')
var atImport          = require("postcss-import")`
`    postcss: [
      pcssCustomProps({preserve: 'computed', strict: false}),
      pcssNext,
      atImport(),
      cssApply,
      pcssNesting,
      pcssReporter({clearMessages: true})
    ]`

**Comment by pascalduez** [comment]
@Memoyr Just a quick remark on what you posted above, `postcss-custom-properties` and `postcss-apply` are already included within `postcss-cssnext`. So apparently you're running them twice.  
Also the plugins order is important, `postcss-import` should come first.

**Comment by pascalduez** [comment]
> I have to import the file into my current file anytime I want to use a custom property set.

That's how `postcss-import` and most preprocessors works.  
Only alt solution I can see for now is waiting for #15.

**Comment by ericsaboia** [comment]
Same is happening here. `apply` can't do its job because something is adding a `\` before the variables. My best guess is that it's `css-loader` doing it, but I can't understand why.

For now to get unblocked I've created [this `postcss` plugin](https://gist.github.com/ericsaboia/eec59a45df48fb981fa5d8d451d8fb22) to clean the extra backslash.

**Comment by pascalduez** [comment]
@ericsaboia Basically `postcss-loader` should comes before the `css-loader` so the `apply` plugin should already have done it's job.

I once tried to debug, but the thing is quite dense, there's so many actors involed... 
Re-opening anyway?

**Comment by ericsaboia** [comment]
thanks for looking into this, even though it doesn't seem to be an issue caused by `postcss-apply`, @pascalduez.

I'm executing `postcss-apply` as a plugin using `extract-text-webpack-plugin` + `optimize-css-assets-webpack-plugin`, so it's only called after all the loaders are executed. 

When I execute it inside the loaders it works well, meaning the `\` isn't there yet. That's why I believe it might be happening inside `css-loader`.

I can set up a minimal reproducible environment to help debug this.

**Comment by yebrahim** [comment]
Also hitting this issue, and also because of `css-loader`. If I switch to the simple `string-loader` it works (although imports break), so it seems `css-loader` does some kind of validation while loading the CSS, and automatically comments invalid CSS out. I can't find a way to turn this off though.

**Comment by pascalduez** [comment]
Okay, I spotted the culprit. This happens when used together with CSS modules. 
It's [`css-selector-tokenizer`](https://github.com/css-modules/css-selector-tokenizer) `stringify` function.

**Comment by yebrahim** [comment]
@pascalduez no way around this then?

**Comment by pascalduez** [comment]
@yebrahim One solution, MR that lib to prevent the escaping. But since custom property sets are not going anywhere, it might be more difficult to get the change accepted, dunno. I'll try, but feel free to.

Currently custom property sets  don't have any browser support and will most likely won't get any.
So it does not make much sense to use the `preserve` option, so they basically won't get escaped (problem solved).
That might be mitigated by the order the plugin are run though.
